### PR TITLE
WIP: Connection management proposal.

### PR DESCRIPTION
**Pull request by brian-js**
[Review on Reviewable](https://reviewable.io/reviews/maidsafe/rfcs/44)

### CODE REVIEW

File: proposed/0008-Connection_management.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+1. In order to prevent any further network activity a terminated state.
+
+```rust
+pub enum Phase {
+    Disconnected(bool),
+    Bootstrapping(bool),
+    Connected(bool),
+    Terminated(bool),
+}
+```
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: Doesn't `Bootstrapping` imply `Disconnected`?  If so, can we remove `Bootstrapping`?

Apart from that, these states are mutually exclusive aren't they?  If so, why do we need the `(bool)` aspect to each - wouldn't a C-style enum do?

File: proposed/0008-Connection_management.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+```rust
+pub struct State {
+    phase: Phase
+}
+```
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: It's not clear why we need to wrap the enum in a struct.  Couldn't we just rename the enum to `State` and hold an instance of that in `RoutingCore`?

File: proposed/0008-Connection_management.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+Both outgoing and incoming connections are created to peers with unknown direction. Use the connect/accept connection code available from crust to store timed state in-line with crust connection handling changes for connections in either direction. This will allow us to determine unambiguously who the connection initiator is and act accordingly.
+
+# Detailed design
+
+## State
+
+Introduce a `State` object to `RoutingCore` representing the distinct phases of execution for network entities characterised as follows.
+
+1. Initially client/node in disconnected state, and when all connections are dropped/lost.
+1. Bootstrapping, initiated in constructor by calling crust service function and halted by crust event sent over channel.
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: Bootstrapping happens whenever the node is disconnected doesn't it, not just in the constructor?

File: proposed/0008-Connection_management.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+- RFC PR: (leave this empty)
+- Issue Number: (leave this empty)
+
+# Summary
+
+Introduce the notion of clearly defined connection phases and establish directed routing network connections in conjunction with crust.
+
+# Motivation
+
+Both outgoing and incoming connections are created to peers with unknown direction. Use the connect/accept connection code available from crust to store timed state in-line with crust connection handling changes for connections in either direction. This will allow us to determine unambiguously who the connection initiator is and act accordingly.
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: I don't think this really explains why we need/want this functionality.  There's no reason given why keeping the state and knowing direction of connections is a good thing or solves an existing issue.

File: proposed/0008-Connection_management.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    peer_addr: SocketAddrW,
+    local_addr: SocketAddrW,
+}
+
+enum ExpectedConnection {
+    InternalRequest::Connect(ConnectRequest),
+    InternalResponse::Connect(ConnectResponse, SignedToken)
+}
+
+pub struct ConnectionFilter<K, V> {
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: Is there a reason for this to be a generic object?

File: proposed/0008-Connection_management.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+```rust
+pub struct State {
+    phase: Phase
+}
+```
+
+## Connection Filter
+
+For connections, add to utils folder a timed `ConnectionFilter` object for key type `crust::Connection`, and value, new type, `ExpectedConnection`. An object of type `ConnectionFilter<Connection, ExpectedConnection>` replaces the current `connection_filter` in `RoutingNode`.
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: I don't think there _is_ a utils folder in Routing.

Comment (Fraser999) [inline proposed/0008-Connection_management.md]: I think we should briefly explain what we mean by a timed filter, either here where we introduce it, or below where we describe its usage.

File: proposed/0008-Connection_management.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+}
+```
+
+```rust
+pub struct State {
+    phase: Phase
+}
+```
+
+## Connection Filter
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: I can't say I like the name "Connection Filter".  I think something like "Expected Connections" better represents it.

File: proposed/0008-Connection_management.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+## Updates to Existing Code
+
+1. Merge the `RoutingNode` functions `handle_new_connection` and `handle_new_bootstrap_connection`.
+1. Remove `Unidentified` connections from `ConnectionName`.
+1. In the event of disconnect implement re-bootstrapping.
+1. Update Hello in line with crust events.
+
+# Drawbacks
+
+N/A
```
Comment (Fraser999) [inline proposed/0008-Connection_management.md]: We've tried this before in the C++ codebase, and I recall problems when two peers connect simultaneously (both are initiators).  We should probably cover that situation in this document.

File: proposed/0008-Connection_management.md (hunk 1 above)
Comment (brian-js) [inline proposed/0008-Connection_management.md]: Yes. After committing it I noticed that.

File: proposed/0008-Connection_management.md (hunk 6 above)
Comment (brian-js) [inline proposed/0008-Connection_management.md]: There will be a utils folder when the implementation gets done. I'll make that clearer.

File: proposed/0008-Connection_management.md (hunk 8 above)
Comment (brian-js) [inline proposed/0008-Connection_management.md]: The idea behind the ExpectedConnections, ConnectionFilter above, is that we will know. I'll try to make the intention clearer before committing without the WIP.

File: proposed/0008-Connection_management.md (hunk 3 above)
Comment (brian-js) [inline proposed/0008-Connection_management.md]: Maybe this needs some clarification on my part. My understanding is that initially we're disconnected. Bootstrapping is initiated in the constructor and stops when crust signals it up or routing calls stop. If we're a node we should enter a connected phase. If we lose all connections we're disconnected and should try bootstrapping again.

File: proposed/0008-Connection_management.md (hunk 4 above)
Comment (brian-js) [inline proposed/0008-Connection_management.md]: Agreed on the explanation.

**Pull request closed without merging**